include/zephyr/arch/posix/arch.h

```diff
@@ -38,6 +40,41 @@ extern "C" {
 #define ARCH_STACK_PTR_ALIGN 4
 #endif
 
+/* "native_posix" should really use CONFIG_LOG_MODE_IMMEDIATE=y but better safe
+ * than sorry: debugging crashes is painful enough already, so try to be nice
+ * and flush all messages. The deferred logging code may also want to enjoy
+ * native_posix too.
+ *
+ * Other archs use Zephyr's elaborate "Fatal Errors" framework which takes care
+ * of flushing logs but native_posix is simpler so we have to do it ourselves.
+ */
+static inline void posix_arch_log_immediate(void)
+{
+#if defined(CONFIG_LOG) && !defined(CONFIG_LOG_MODE_MINIMAL)
+	/* We can't #include the higher-level "log_ctrl.h" in this low-level
+	 * file here because this descends into .h dependency hell.  So let's
+	 * use a one-line forward declaration instead. This void->void
+	 * prototype does not look like it's going to change much in the
+	 * future.
+	 *
+	 * We can't declare "void log_panic(void);" because it is not what it
+	 * seems but a complicated static inline + macro override in
+	 * generated/syscalls/log_ctrl.h. In native_posix mode log_panic()
+	 * boils down to just z_impl_log_panic() so exceptionally invoke that
+	 * directly.
```

Comment:
The second paragraph is true, but can be condensed to something like "ARCH_EXCEPT() can't possibly be used in user mode, and the syscall headers don't work this early".